// @nostrify/nostrify/dist/NBrowserSigner.js
class NBrowserSigner {
  get nostr() {
    const nostr = globalThis.nostr;
    if (!nostr) {
      throw new Error("Browser extension not available");
    }
    return nostr;
  }
  async getPublicKey() {
    const pubkey = await this.nostr.getPublicKey();
    if (typeof pubkey !== "string") {
      throw new Error(`Nostr public key retrieval failed: expected string, got ${JSON.stringify(pubkey)}`);
    }
    return pubkey;
  }
  async signEvent(event) {
    const signed = await this.nostr.signEvent(event);
    if (typeof signed !== "object" || !signed.id || !signed.pubkey || !signed.sig) {
      throw new Error(
        `Nostr event signing failed: expected object with id, pubkey, and sig, got ${JSON.stringify(signed)}`
      );
    }
    return signed;
  }
  async getRelays() {
    if (!this.nostr.getRelays) {
      return {};
    }
    const relays = await this.nostr.getRelays();
    if (typeof relays !== "object" || relays === null) {
      throw new Error(`Nostr getRelays failed: expected object, got ${JSON.stringify(relays)}`);
    }
    return this.nostr.getRelays();
  }
  get nip04() {
    const nostr = this.nostr;
    if (!nostr.nip04) {
      return void 0;
    }
    return {
      encrypt: async (pubkey, plaintext) => {
        const encrypted = await nostr.nip04.encrypt(pubkey, plaintext);
        if (typeof encrypted !== "string") {
          throw new Error(`NIP-04 encryption failed: expected string result, got ${JSON.stringify(encrypted)}`);
        }
        return encrypted;
      },
      decrypt: async (pubkey, ciphertext) => {
        const decrypted = await nostr.nip04.decrypt(pubkey, ciphertext);
        if (typeof decrypted !== "string") {
          throw new Error(`NIP-04 decryption failed: expected string result, got ${JSON.stringify(decrypted)}`);
        }
        return decrypted;
      }
    };
  }
  get nip44() {
    const nostr = this.nostr;
    if (!nostr.nip44) {
      return void 0;
    }
    return {
      encrypt: async (pubkey, plaintext) => {
        const encrypted = await nostr.nip44.encrypt(pubkey, plaintext);
        if (typeof encrypted !== "string") {
          throw new Error(`NIP-44 encryption failed: expected string result, got ${JSON.stringify(encrypted)}`);
        }
        return encrypted;
      },
      decrypt: async (pubkey, ciphertext) => {
        const decrypted = await nostr.nip44.decrypt(pubkey, ciphertext);
        if (typeof decrypted !== "string") {
          throw new Error(`NIP-44 decryption failed: expected string result, got ${JSON.stringify(decrypted)}`);
        }
        return decrypted;
      }
    };
  }
}
export {
  NBrowserSigner
};
