// @nostrify/nostrify/dist/NKinds.js
class NKinds {
  /** Events are **regular**, which means they're all expected to be stored by relays. */
  static regular(kind) {
    return 1e3 <= kind && kind < 1e4 || [1, 2, 4, 5, 6, 7, 8, 16, 40, 41, 42, 43, 44].includes(kind);
  }
  /** Events are **replaceable**, which means that, for each combination of `pubkey` and `kind`, only the latest event is expected to (SHOULD) be stored by relays, older versions are expected to be discarded. */
  static replaceable(kind) {
    return 1e4 <= kind && kind < 2e4 || [0, 3].includes(kind);
  }
  /** Events are **ephemeral**, which means they are not expected to be stored by relays. */
  static ephemeral(kind) {
    return 2e4 <= kind && kind < 3e4;
  }
  /** Events are **addressable**, which means that, for each combination of `pubkey`, `kind` and the `d` tag, only the latest event is expected to be stored by relays, older versions are expected to be discarded. */
  static addressable(kind) {
    return 3e4 <= kind && kind < 4e4;
  }
  /** @deprecated Use `NKinds.addressable()` instead. */
  static parameterizedReplaceable(kind) {
    return NKinds.addressable(kind);
  }
}
export {
  NKinds
};
