// @nostrify/nostrify/dist/test/TestRelayServer.js
import { MockRelay } from "./mod.js";
import { NSchema as n } from "../NSchema.js";
import { WebSocket, WebSocketServer } from "ws";
import { createServer } from "node:http";
class TestRelayServer {
  port = 0;
  inited = false;
  httpServer;
  wsServer;
  opts;
  connections = /* @__PURE__ */ new Set();
  controllers = /* @__PURE__ */ new Map();
  store = new MockRelay();
  constructor(opts) {
    this.opts = opts || {};
    this.httpServer = createServer();
    this.wsServer = new WebSocketServer({ server: this.httpServer });
    this.setupWebSocketServer();
  }
  init() {
    const { resolve, promise } = Promise.withResolvers();
    this.httpServer.listen(0, "127.0.0.1", () => {
      this.port = this.httpServer.address().port;
      this.inited = true;
      resolve();
    });
    return promise;
  }
  setupWebSocketServer() {
    this.wsServer.on("connection", (socket) => {
      this.connections.add(socket);
      socket.on("close", () => {
        this.connections.delete(socket);
        for (const [subId, controller] of this.controllers.entries()) {
          controller.abort();
          this.controllers.delete(subId);
        }
      });
      socket.on("message", (data) => {
        try {
          const result = n.json().pipe(n.clientMsg()).safeParse(data.toString());
          if (result.success) {
            const handleMessage = this.opts?.handleMessage ?? this.handleMessage.bind(this);
            handleMessage(socket, result.data);
          }
        } catch {
        }
      });
      socket.on("error", (error) => {
        console.error("WebSocket error:", error);
        this.connections.delete(socket);
      });
    });
  }
  send(socket, msg) {
    if (!this.inited) throw new Error("TestRelayServer not initialized");
    if (socket.readyState === WebSocket.OPEN) {
      socket.send(JSON.stringify(msg));
    }
  }
  async handleMessage(socket, msg) {
    if (!this.inited) throw new Error("TestRelayServer not initialized");
    switch (msg[0]) {
      case "REQ": {
        const [_, subId, ...filters] = msg;
        const controller = new AbortController();
        this.controllers.set(subId, controller);
        try {
          for await (const msg2 of this.store.req(filters, { signal: controller.signal })) {
            msg2[1] = subId;
            this.send(socket, msg2);
          }
        } catch {
        }
        break;
      }
      case "CLOSE": {
        const subId = msg[1];
        this.controllers.get(subId)?.abort();
        this.controllers.delete(subId);
        break;
      }
      case "EVENT": {
        const [_, event] = msg;
        await this.store.event(event);
        this.send(socket, ["OK", event.id, true, ""]);
        break;
      }
    }
  }
  get url() {
    if (!this.inited) throw new Error("TestRelayServer not initialized");
    const addr = this.httpServer.address();
    return `ws://${addr.address}:${addr.port}`;
  }
  // deno-lint-ignore require-await
  async close() {
    if (!this.inited) throw new Error("TestRelayServer not initialized");
    return new Promise((resolve) => {
      this.connections.forEach((conn) => {
        if (conn.readyState === WebSocket.OPEN) {
          conn.close();
        }
      });
      this.connections.clear();
      this.controllers.forEach((controller) => controller.abort());
      this.controllers.clear();
      this.wsServer.close(() => {
        this.httpServer.close(() => {
          this.inited = false;
          resolve();
        });
      });
    });
  }
  open() {
    if (this.inited) throw new Error("TestRelayServer already initialized");
    if (!this.httpServer.listening) {
      const { resolve, promise } = Promise.withResolvers();
      this.httpServer = createServer();
      this.wsServer = new WebSocketServer({ server: this.httpServer });
      this.setupWebSocketServer();
      this.httpServer.listen(0, "127.0.0.1", () => {
        this.port = this.httpServer.address().port;
        this.inited = true;
        resolve();
      });
      return promise;
    }
    return Promise.resolve();
  }
  event(event) {
    if (!this.inited) throw new Error("TestRelayServer not initialized");
    return this.store.event(event);
  }
  async [Symbol.asyncDispose]() {
    if (this.inited) {
      await this.close();
    }
  }
  static async create(opts) {
    const server = new TestRelayServer(opts);
    await server.init();
    return server;
  }
}
export {
  TestRelayServer
};
